File: test/TestPatient/test_patient.cpp
#include <iostream>
#include <gtest/gtest.h>
#include "../../Patient/Patient.cpp"

using namespace std;

TEST(CreatePatientObject, create_default_patient)
{
    string expected_name = "";
    string expected_gender = "";
    int expected_age = 0;
    string expected_procedure = "";

    Patient default_patient;

    EXPECT_STREQ(expected_name.c_str(), default_patient.getName().c_str());
    EXPECT_STREQ(expected_gender.c_str(), default_patient.getGender().c_str());
    EXPECT_STREQ(expected_procedure.c_str(), default_patient.getProcedureName().c_str());
    EXPECT_EQ(expected_age, default_patient.getAge());
}


TEST(CreatePatientObject, create_custom_patient)
{
    string expected_name = "XYZ";
    string expected_gender = "Male";
    int expected_age = 20;
    string expected_procedure = "Cardiac";

    Patient custom_patient("XYZ", "Male", 20, "Cardiac");

    EXPECT_STREQ(expected_name.c_str(), custom_patient.getName().c_str());
    EXPECT_STREQ(expected_gender.c_str(), custom_patient.getGender().c_str());
    EXPECT_STREQ(expected_procedure.c_str(), custom_patient.getProcedureName().c_str());
    EXPECT_EQ(expected_age, custom_patient.getAge());
}


TEST(APITest, test_getters_and_setters)
{
    Patient patient;
    patient.setName("ABC");
    patient.setAge(20);
    patient.setGender("Male");
    patient.setProcedureName("Cardiac");
    patient.setBusyStatus(1);

    EXPECT_STREQ("ABC", patient.getName().c_str());
    EXPECT_STREQ("Male", patient.getGender().c_str());
    EXPECT_STREQ("Cardiac", patient.getProcedureName().c_str());
    EXPECT_EQ(20, patient.getAge());
    EXPECT_EQ(1,patient.getBusyStatus());
}

TEST(APITest, test_to_string_method)
{
    Patient patient;
    patient.setName("ABC");
    patient.setAge(20);
    patient.setGender("Male");
    patient.setProcedureName("Cardiac");
    string expected_string = "0;ABC;Male;20;Cardiac;;;0";
    EXPECT_STREQ(expected_string.c_str(), patient.toString().c_str());
}

TEST(APITest, test_add_single_report_id)
{
    Patient patient;
    EXPECT_STREQ("", patient.getReportIds().c_str());
    patient.addReportId("12");
    EXPECT_STREQ("12", patient.getReportIds().c_str());
}

TEST(APITest, test_add_single_consumable)
{
    Patient patient;
    EXPECT_STREQ("", patient.getConsumables().c_str());
    patient.addConsumable("Gloves");
    EXPECT_STREQ("Gloves", patient.getConsumables().c_str());
}

TEST(APITest, test_add_multiple_report_ids)
{
    Patient patient;
    patient.addReportId("11");
    EXPECT_STREQ("11", patient.getReportIds().c_str());
    patient.addReportId("12");
    EXPECT_STREQ("11,12", patient.getReportIds().c_str());
    patient.addReportId("13");
    EXPECT_STREQ("11,12,13", patient.getReportIds().c_str());
}

TEST(APITest, test_add_multiple_consumables)
{
    Patient patient;
    patient.addConsumable("Gloves");
    EXPECT_STREQ("Gloves", patient.getConsumables().c_str());
    patient.addConsumable("Mask");
    EXPECT_STREQ("Gloves,Mask", patient.getConsumables().c_str());
}


TEST(Operator_Test, Assignment_Operator)
{
    Patient p1("ABC", "Male", 20, "Cardiac");
    Patient p2;
    p2 = p1;
    EXPECT_STREQ(p1.getName().c_str(), p2.getName().c_str());
    EXPECT_STREQ(p1.getGender().c_str(), p2.getGender().c_str());
    EXPECT_STREQ(p1.getProcedureName().c_str(), p2.getProcedureName().c_str());
    EXPECT_EQ(p1.getAge(), p2.getAge());
}

int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: test/PythonUnitTests/tests/test_patient.py
import unittest
from PythonClient.patient import Patient

class TestPatient(unittest.TestCase):

    def test_create_default_patient(self):
        p = Patient()
        self.assertEqual(p.getName(), "")
        self.assertEqual(p.getAge(), 0)
        self.assertEqual(p.getGender(), "")
        self.assertEqual(p.getProcedureName(), "")
        self.assertEqual(p.getConsumables(), "")
        self.assertEqual(p.getReportIds(), "")
        self.assertFalse(p.getBusyStatus())
    

    def test_create_patient(self):
        p = Patient("ABC", 30, "Male", "Cardiac")
        self.assertEqual(p.getName(), "ABC")
        self.assertEqual(p.getAge(), 30)
        self.assertEqual(p.getGender(), "Male")
        self.assertEqual(p.getProcedureName(), "Cardiac")
    

    def test_setters_and_getters(self):
        p = Patient()
        p.setId("ABC123")
        p.setName("ABC")
        p.setAge(25)
        p.setGender("Female")
        p.setProcedureName("Cardiac")
        p.setReportIds("ECG123")
        p.setConsumables("Catheter")
        p.setBusyStatus(1)

        self.assertEqual(p.getId(), "ABC123")
        self.assertEqual(p.getName(), "ABC")
        self.assertEqual(p.getAge(), 25)
        self.assertEqual(p.getGender(), "Female")
        self.assertEqual(p.getProcedureName(), "Cardiac")
        self.assertEqual(p.getReportIds(), "ECG123")
        self.assertEqual(p.getConsumables(), "Catheter")
        self.assertEqual(p.getBusyStatus(), 1)


    def test_add_multiple_consumables(self):
        p = Patient()
        p.addConsumables("Gloves")
        p.addConsumables("Mask")

        self.assertEqual(p.getConsumables(), "Gloves,Mask")
    
    def test_add_multiple_report_ids(self):
        p = Patient()
        p.addReportIds("ECG123")
        p.addReportIds("ECG124")

        self.assertEqual(p.getReportIds(), "ECG123,ECG124")
    

    def test_to_string_method(self):
        p = Patient()
        p.setName("ABC")
        p.setAge(25)
        p.setGender("Female")
        p.setProcedureName("Cardiac")
        p.setReportIds("ECG123")
        p.setConsumables("Catheter")

        expected_string = ";ABC;Female;25;Cardiac;Catheter;ECG123;0"
        self.assertEqual(p.toString(), expected_string)
    

    def test_equality_operator(self):
        p1 = Patient("ABC", 25, "Male", "Xray")
        p2 = Patient("ABC", 25, "Male", "Xray")
        p3 = Patient()

        self.assertTrue(p1 == p2)
        self.assertFalse(p2 == p3)


    def test_pateint_repr(self):
        p = Patient()
        p.setName("ABC")
        p.setAge(25)
        p.setGender("Female")
        p.setProcedureName("Cardiac")
        p.setReportIds("ECG123")
        p.setConsumables("Catheter")
        expected_string = ";ABC;Female;25;Cardiac;Catheter;ECG123;0"

        self.assertEqual(repr(p), expected_string)
